File: idp-pipelines/catalog-fed-ramp/fips/go/fips_check.go
package main

import (
	"crypto/hmac"
	"crypto/sha256"
	"fmt"
	"io/ioutil"
	"os"
	"os/exec"
	"strings"
)

// Check if FIPS mode is enabled on the system
func checkFipsMode() error {
	// First, check /proc/sys/crypto/fips_enabled
	data, err := ioutil.ReadFile("/proc/sys/crypto/fips_enabled")
	if err == nil {
		if strings.TrimSpace(string(data)) == "1" {
			fmt.Println("FIPS mode is enabled.")
			return nil
		} else {
			fmt.Println("FIPS mode is not enabled.")
			return fmt.Errorf("fips mode is not enabled")
		}
	}

	// If the file does not exist, fall back to OpenSSL check
	fmt.Println("Checking OpenSSL for FIPS mode...")
	cmd := exec.Command("openssl", "fips")
	output, err := cmd.CombinedOutput()
	if err != nil || !strings.Contains(string(output), "FIPS mode enabled") {
		fmt.Println("FIPS mode is NOT enabled (OpenSSL).")
		return fmt.Errorf("fips mode is not enabled (OpenSSL)")
	}

	fmt.Println("FIPS mode is enabled (OpenSSL).")
	return nil
}

// Check FIPS-compliant SHA-256 and HMAC using Go's crypto package
func checkFipsSha256() bool {
	// Hashing with SHA-256 (FIPS-compliant)
	data := []byte("Test data")
	hash := sha256.New()
	hash.Write(data)
	digest := hash.Sum(nil)
	fmt.Printf("SHA-256 hash: %x\n", digest)

	// Using HMAC with SHA-256 (FIPS-compliant)
	key := []byte("secret")
	hmac := hmac.New(sha256.New, key)
	hmac.Write(data)
	hmacDigest := hmac.Sum(nil)
	fmt.Printf("HMAC-SHA-256: %x\n", hmacDigest)

	// Return true if both are non-nil
	if digest != nil && hmacDigest != nil {
		return true
	}
	return false
}

func main() {


	fmt.Println("Testing go")

	// Check FIPS-compliant hashing and HMAC (SHA-256)
	if checkFipsSha256() {
		fmt.Println("SHA-256 and HMAC are FIPS-compliant.")
	} else {
		fmt.Println("FIPS-compliant hashing or HMAC failed.")
		os.Exit(1)
	}


	// Check FIPS mode
	//err := checkFipsMode()
	//if err != nil {
	//		os.Exit(1)
	//}


	// Successfully passed both checks
	os.Exit(0)
}

File: idp-pipelines/catalog-fed-ramp/fips/nodejs/fips_check.js
const crypto = require('crypto');
const { exec } = require('child_process');

// Function to check SHA-256 and HMAC-SHA-256, verifying FIPS compliance
function checkFipsSha256() {
    const data = Buffer.from('Test data');

    // SHA-256 hash using crypto module
    const sha256Hash = crypto.createHash('sha256');
    sha256Hash.update(data);
    const sha256Value = sha256Hash.digest('hex');
    console.log(`SHA-256 hash: ${sha256Value}`);

    // HMAC with SHA-256 using crypto module (FIPS-compliant if in FIPS mode)
    const key = Buffer.from('secret');
    const hmac = crypto.createHmac('sha256', key);
    hmac.update(data);
    const hmacValue = hmac.digest('hex');
    console.log(`HMAC-SHA-256: ${hmacValue}`);

    // Assuming the check is successful if no errors occurred
    return sha256Value && hmacValue ? true : false;
}

// Function to check if FIPS mode is enabled on the system
function checkFipsMode() {
    exec('cat /proc/sys/crypto/fips_enabled', (err, stdout, stderr) => {
        if (err) {
            console.log('Error checking FIPS mode: ', err);
            // Fallback to OpenSSL check if the above fails
            exec('openssl fips', (opensslErr, opensslStdout, opensslStderr) => {
                if (opensslErr) {
                    console.log('Error checking OpenSSL FIPS mode: ', opensslErr);
                } else if (opensslStdout.includes('FIPS mode enabled')) {
                    console.log('FIPS mode is enabled (OpenSSL).');
                } else {
                    console.log('FIPS mode is NOT enabled (OpenSSL).');
                }
            });
        } else {
            if (stdout.trim() === '1') {
                console.log('FIPS mode is enabled.');
            } else {
                console.log('FIPS mode is not enabled.');
            }
        }
    });
}

// Main function to run both checks
function runChecks() {
    try {
        console.log('FIPS-compliant Testing nodej');
        const sha256Check = checkFipsSha256();
        if (sha256Check) {
            console.log('FIPS-compliant SHA-256 and HMAC-SHA-256 operations passed.');
        } else {
            console.log('FIPS-compliant operations failed.');
        }
        //checkFipsMode();
    } catch (e) {
        console.error('An error occurred: ', e);
        process.exit(1);
    }
}

// Run the checks
runChecks();
